tests/integration/test_github_workflows.py

```
@@ -1888,4 +1923,862 @@
+            if 'permissions' in workflow_content:
+                perms = workflow_content['permissions']
+                
+                # If permissions are defined, they should be specific
+                if isinstance(perms, dict):
+                    # Should not have write-all
+                    assert perms.get('contents') != 'write' or \
+                           perms.get('pull-requests') == 'write', \
+                        f"Overly permissive token in {workflow_path}"
+
+
+class TestRequirementsDevValidation:
```

Ruff (0.14.6)

Redefinition of unused `TestRequirementsDevValidation` from line 2283

(F811)